tesseract_core/sdk/engine.py

```diff
@@ -116,14 +118,31 @@ def wrapper_needs_docker(*args: Any, **kwargs: Any) -> None:
     return wrapper_needs_docker
 
 
-def get_free_port() -> int:
-    """Find a free port to use for HTTP."""
-    import socket
-    from contextlib import closing
+def get_free_port(within_range: tuple[int, int] | None = None) -> int:
+    """Find a random free port to use for HTTP."""
+    if within_range is None:
+        # Let OS pick a random free port
+        with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+            s.bind(("localhost", 0))
+            return s.getsockname()[1]
 
-    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
-        s.bind(("localhost", 0))
-        return s.getsockname()[1]
+    start, end = within_range
+    if start < 0 or end > 65535 or start >= end:
+        raise ValueError("Invalid port range")
+
+    # Try random ports in the given range
+    portlist = list(range(start, end))
+    random.shuffle(portlist)
```

Comment:
Why shuffle the list?

Reply:
Mostly to emulate the previous behavior (which was effectively randomized), but also to avoid trying a lot of ports when certain port ranges are blocked or taken. Call it defensive programming if you will.